src/consensus/aft/raft.h

```diff
@@ -380,20 +380,11 @@ namespace aft
     std::optional<kv::Consensus::SignableTxIndices> get_signable_txid() override
     {
       std::lock_guard<std::mutex> guard(state->lock);
-      if (
-        consensus_type == ConsensusType::BFT ||
-        state->commit_idx >= election_index)
-      {
-        kv::Consensus::SignableTxIndices r;
-        r.term = get_term_internal(state->commit_idx);
-        r.version = state->commit_idx;
-        r.previous_version = last_committable_index();
-        return r;
-      }
-      else
-      {
-        return std::nullopt;
-      }
+      kv::Consensus::SignableTxIndices r;
+      r.version = state->last_idx;
+      r.term = get_term_internal(r.version);
+      r.previous_version = std::max(last_committable_index(), election_index);
+      return r;
```

Comment:
Probably also should remove the optional wrapper, now this is always filled. This was used to prevent a signature being emitted in the short period after an election where the new primary doesn't know commit, to avoid signatures that say commit moved backwards. That's no longer possible - the new primary _must_ produce a new signature, and that's the first thing they _could_ commit. `previous_version` is used to truncate the merkle tree, so needs to be reasonably fresh - it's possible this is still insufficient.

Reply:
That sounds about right, but I am tempted to add a check to read_ledger to confirm the trees we produce always cover the right interval. `previous_version` is only used to select the range for the serialisation, the compaction is done on commit.

It's worth noting that the first signature of every term will now report that the commit has reached the end of the previous term, which is not necessarily true. We know we need to remove those fields in any case.

Reply:
Will add this check in a separate PR, as discussed.